### PR TITLE
subscribe: add subscribe service for streaming change events

### PR DESCRIPTION
**Pull request by dnephin**
Branched from #8711, and extracted from 1d0f3c4853f8e125c85ee67d58965b59d0246ff9

This PR adds an implementation of `pbsubscribe.StateChangeSubscriptionServer` and adds the server to the grpcHandler used by the rpc server.

The tests are quite verbose and might benefit from some additional helpers to make them more readable.

### CODE REVIEW

File: agent/subscribe/subscribe.go
Comment (banks) [inline agent/subscribe/subscribe.go:131]: When ACLs are disabled IIRC? It could just be defensive but we certainly have places in Endpoint handlers where authz resolves to nil if ACLs are disabled.

**Review by banks — COMMENTED** [review]
This looks really good overall.

My one question is whether `agent/subscribe` is the best location for this new package. It's a server-only component as such the most natural place by existing conventions would be in `agent/consul` where all the other server-only code is.

Was putting it in `agent` dir intentional for some reason?

File: agent/subscribe/subscribe.go
Comment (dnephin) [reply agent/subscribe/subscribe.go:131]: aha, yes, that is the case. I guess I scanned over that code too quickly. I'll update this to a comment which says so

**Comment by dnephin** [comment]
> My one question is whether `agent/subscribe` is the best location for this new package. It's a server-only component as such the most natural place by existing conventions would be in `agent/consul` where all the other server-only code is.

My thinking is that repeating the project name in the import path doesn't help the reader very much. It doesn't provide any additional context to help understand the contents of the package. Once one is familiar with consul they may do the mental mapping of `consul` -> `server`, but even then it's not exactly right. At the very least `client{_serf}.go` in that package are for the client. 

While this package is most likely only ever going to be used by the consul Server process, there is nothing really inherently tied to the Servers (compare to components like fsm/raft/state). Anything that can implement the `subscribe.Backend` could use this package, for example to proxy requests to a real Server. I guess that is pretty unlikely to ever happen.

I do see your point about the `agent` sub-packages being primarily either client-specific, or at least used by both client and server. 

Maybe `agent/rpc/subscribe` would be a better place? In both #8580 and an upcoming streaming PR I am proposing we add new packages to `agent/rpclient/` which handle the RPC calls along with caching and any switching between implementations (ex: grpc vs net/rpc, or blocking vs streaming). So `agent/rpc/subscribe` would mirror that nicely.

**Comment by dnephin** [comment]
The latest commit moves `agent/subscribe` to `agent/rpc/subscribe`. Thankfully with GoLand it only takes 5 seconds to move a packages.

File: agent/rpc/subscribe/auth.go
Comment (dnephin) [inline agent/rpc/subscribe/auth.go:24]: Calling out this TODO. Should this code expect to handle `acl.Default`? I guess currently it assumes default deny is ACLs are enabled, which I think I've heard is the expectation, but I thought I should double check.

**Comment by mkeeler** [comment]
I think it may already be sorted but while its on my mind I would say that we should move away from server code being in the `consul` package. Really that package is bloated and we need to modularize more of the functionality in that package and place them into other packages elsewhere in the tree.

File: agent/rpc/subscribe/auth.go
Comment (mkeeler) [reply agent/rpc/subscribe/auth.go:24]: `acl.Default` is to allow chaining Authorizers and delegating the decision to something further on in the chain rather than making an explicit allow/deny decision. The `acl.ChainedAuthorizer` type is what Consul creates and it specially processes the `acl.Default` decision from a child authorizer and proceeds with the next authorizer in the chain.

https://github.com/hashicorp/consul/blob/bf2b52f88061520c4c5b996df91d40023537e46b/acl/chained_authorizer.go#L26-L34

Comment (mkeeler) [reply agent/rpc/subscribe/auth.go:24]: So basically the top level authorizer should return `Allow` or things should be denied.

File: agent/grpc/handler.go
Comment (mkeeler) [inline agent/grpc/handler.go:23]: It might be due to a lack of context on my part but I don't understand why this function should execute the register func instead of allowing the caller a way to register after creation.

Wouldn't it be cleaner to just export the `Srv` field in the `Handler` and allow the caller of this function to register endpoints with the server after the fact.

File: agent/rpc/subscribe/auth.go
Comment (mkeeler) [reply agent/rpc/subscribe/auth.go:24]: Also as far as the ACL package is concerned the true default decision is always `Deny`. From the overall product perspective the `ChainedAuthorizer` has to be created with the final authorizer in the chain as a `StaticAuthorizer` set to implement the products `default_policy = "allow"` mode.

File: agent/rpc/subscribe/auth.go
Comment (mkeeler) [inline agent/rpc/subscribe/auth.go:33]:
```suggestion
		var authzCtx acl.AuthorizerContext
		csn.Service.FillAuthzContext(&authCtx)

		if dec := authz.NodeRead(csn.Node.Node, &authzCtx); dec != acl.Allow {
			return acl.Deny
		}

		// Enterprise support for streaming events - they don't have enough data to
		// populate it yet.
		if dec := authz.ServiceRead(csn.Service.Service, &authzCtx); dec != acl.Allow {
```

I am not entirely sure yet whether the underlying types are the same and this will actually work but this is generally how its done in other areas.

File: agent/rpc/subscribe/logger.go
Comment (mkeeler) [inline agent/rpc/subscribe/logger.go:44]: What is the `Key`. It sounds similar to the cache key which eventually contains an ACL token secret and we would not want to log those secrets. My hunch is that this is different but it would be great for you to verify that.

File: agent/rpc/subscribe/subscribe.go
Comment (mkeeler) [inline agent/rpc/subscribe/subscribe.go:38]: Does streaming need to perform ACL token based namespace inference?

In many other places we call the other `ResolveTokenAndDefaultMeta` to perform the `EnterpriseMeta` defaulting based on the token and my guess is that we will want subscriptions to also do this defaulting.

File: agent/rpc/subscribe/subscribe.go
Comment (mkeeler) [inline agent/rpc/subscribe/subscribe.go:54]: Do you mean a token expiring during the stream or being expired at this point in the code. Because if a token is actually expired right here ResolveToken should return some sort of ACL error.

In the other case of being valid now but expiring later you should periodically re-resolve the token. For the xDS server we have a ticker which fires every so often to force authorization of the stream to happen again and it seems like a similar change could be made here as a first step.

File: agent/rpc/subscribe/subscribe.go
Comment (mkeeler) [inline agent/rpc/subscribe/subscribe.go:168]: Why not handle these cases in switch within `setPayload`?

File: agent/grpc/handler.go
Comment (dnephin) [reply agent/grpc/handler.go:23]: Exposing the field could also work, and I think ends up being not all that much different, since the caller could technically capture the srv using the register function (as one test in this PR does).

I like the register func approach because I think it makes the usage more obvious. From the function definition you can see that it requires a register function to add services, and the return type has 3 exported methods that can be used.

If the returned type expose `Srv` it's less clear to the caller what needs to be done. They could call `Handler.Srv.Serve()` directly with the wrong listener, which would cause `Handle()` to block and never unblock. 

Ideally I would have had the `register` function take an interface with only the `RegisterService` call on it, but unfortunately the gRPC `RegisterXServer` functions accept the concrete type and not an interface, so that isn't an option. (It could actually work if we use the `_X_serviceDesc` vars that gRPC exposes, but the naming of those makes me believe they are probably intended to be implementation details and may change when we upgrade).

File: agent/rpc/subscribe/auth.go
Comment (dnephin) [reply agent/rpc/subscribe/auth.go:33]: Thanks! This code seems to work. I've been wondering if there isn't already some code somewhere that does this authorization for a CheckServiceNode. It might make sense to have them use the same function

File: agent/rpc/subscribe/logger.go
Comment (dnephin) [reply agent/rpc/subscribe/logger.go:44]: It is this: https://github.com/hashicorp/consul/blob/master/proto/pbsubscribe/subscribe.proto#L46-L51, no secrets in this key.

File: agent/rpc/subscribe/subscribe.go
Comment (dnephin) [reply agent/rpc/subscribe/subscribe.go:38]: I have no idea, but I guess so, if that's what we do for other RPC endpoints. I'll change this interface to `ResolveTokenAndDefaultMeta`

File: agent/rpc/subscribe/subscribe.go
Comment (dnephin) [reply agent/rpc/subscribe/subscribe.go:54]: The way it works right now (https://github.com/hashicorp/consul/blob/master/agent/consul/state/acl_events.go) is that if any of the ACLs change we build up a list of tokens that are impacted, create a special event that is handled by EventPublisher, and it closes any of the subscription which used those tokens, with a specific error.

This TODO came from the original diff, but I left it because I think the idea is that force closing all the impacted subscriptions is not "graceful". I don't think it's something we need to fix for the first release.

File: agent/rpc/subscribe/subscribe.go
Comment (dnephin) [reply agent/rpc/subscribe/subscribe.go:168]: I like that idea. I don't think it addresses the TODO, but it should remove one extra switch

**Comment by dnephin** [comment]
I'm going to start a new branch based on this one and work on the batch event and acl changes

File: agent/rpc/subscribe/subscribe.go
Comment (dnephin) [reply agent/rpc/subscribe/subscribe.go:168]: I tried it out. The reason reason this doesn't work is that the types for the framing events are not exported. So we need separate switches either way. One is a type switch and the other uses no clause and the cases themselves evaluate to true/false.

File: agent/grpc/handler.go
Comment (mkeeler) [reply agent/grpc/handler.go:23]: I suspected it might be a dependent library as the reason for not being able just use a func on the Handler itself. In that case this is fine.

File: agent/rpc/subscribe/auth.go
Comment (mkeeler) [reply agent/rpc/subscribe/auth.go:33]: I don't believe a common check exists anywhere. There are a number of acl helpers in agent/acl.go but none for this use case. I think having methods on the types for validating read/write access would be good for the CheckServiceNode type and many others. Similar to what we have for config entries:

https://github.com/hashicorp/consul/blob/364f6589c89f146b1617842b03c7c4ed2227b157/agent/structs/config_entry.go#L42-L45

None of that exists though.

File: agent/rpc/subscribe/subscribe.go
Comment (mkeeler) [reply agent/rpc/subscribe/subscribe.go:168]: Sounds good to me.

File: agent/rpc/subscribe/subscribe.go
Comment (dnephin) [reply agent/rpc/subscribe/subscribe.go:38]: I don't fully understand what this does. In what cases would we need to perform ACL token based namespace inference, vs not? The token can have a default namespace, if one isn't set in the request?

The `SubscriptionRequest` doesn't have an EnterpriseMeta.Namespace yet, so if the above is correct, I think this be done when adding the Namespace to SubscriptionRequest, which should be one of the next follow up PRs.

File: agent/rpc/subscribe/auth.go
Comment (dnephin) [reply agent/rpc/subscribe/auth.go:33]: Nice! I can add that.  I notice these do not take an `AuthorzierContext`, even though one would need to be created for `ResolveTokenAndDefaultMeta`. Is the context that is passed to `ResolveTokenAndDefaultMeta` captured by the authorizer, so that it doesn't have to be passed to these `CanRead`/`CanWrite` ?

Comment (mkeeler) [reply agent/rpc/subscribe/auth.go]: For Config Entries and plenty of other data, all the information to populate the authz context with is stored in the object so the authz context is created internally.

https://github.com/hashicorp/consul/blob/0c9177f6a5ddd5eb9cf58a36ea9f98f56197c8c4/agent/structs/config_entry.go#L136-L140

Comment (dnephin) [reply agent/rpc/subscribe/auth.go]: Ahh, I had completed missed the `FillAuthzContext` line. Thanks! This makes sense now.

**Review by kyhavlov — APPROVED** [review]
LGTM! Discussed the remaining TODOs and it makes sense to cover the important ones in follow-up PRs.

**Pull request merged**
